Pal/lib/slabmgr.h

```diff
@@ -301,16 +303,39 @@ static inline int enlarge_slab_mgr (SLAB_MGR mgr, int level)
      * every time it grows.  The assumption if we get this far is that
      * mgr->addr == mgr->top_addr */
     assert(mgr->addr[level] == mgr->addr_top[level]);
-    
+
     size_t size = mgr->size[level];
-    SLAB_AREA area = (SLAB_AREA) system_malloc(__MAX_MEM_SIZE(slab_levels[level], size));
-    if (!area)
+    SLAB_AREA area;
+
+    /* If there is a previously allocated area, just activate it. */
+    area = listp_prev_entry(mgr->active_area[level], &mgr->area_list[level], __list);
+    if (area) {
+        __set_free_slab_area(area, mgr, level);
+        return 0;
+    }
+
+    /* system_malloc() may be blocking, so we release the lock before
+     * allocating more memory */
+    system_unlock();
+
+    area = (SLAB_AREA) system_malloc(__MAX_MEM_SIZE(slab_levels[level], size));
```

Comment:
I don't like this convention of releasing a lock "just for a moment" and acquiring it again. The caller of this function can't expect this and it may lead to hard-to-spot race conditions.

Reply:
It will be a problem if the function did not grab the lock itself. However, when using this function, the caller expects the function to maintain atomicity, and we have done our best to achieve atomicity.

Reply:
> It will be a problem if the function did not grab the lock itself.

I didn't mean the caller forgetting to take the lock. The problem is that the calling code may look like this:
```C
if (some_condition_on_mgr_object_holds) {
    if (no_more_free_space)
        enlarge_slab_mgr(mgr, level);
    // Do something which relies on `some_condition_on_mgr_object_holds`.
}
```
The race-condition here is almost impossible to spot, because the reader wouldn't assume that the lock this function acquired is released for some time inside `enlarge_slab_mgr`.

Reply:
No, in that case, the programmer made a mistake by not protecting this critical section with its own lock. If the programmer reuses the same lock that he/she assigns to the memory collector, this code wouldn't even work because of deadlock.

I don't believe there is a race condition here except the potential memory draining condition you mentioned. Again, that only happens in extremely rare condition.

Reply:
The caller code is most likely to be inside slabmgr.h, which uses the same lock for everything.
Please at least add a comment right before `enlarge_slab_mgr` definition warning about this behavior.

Reply:
In general, I will come down on the side of: it is hard, if not impossible, to have too many comments about how things are synchronized.

Reply:
@donporter Yup, but IMO this is a special case, which can't be expected by a caller/code reviewer, so IMO it deserves at least a short comment.

Other solution to such a problem could be to require the caller to release the lock before calling and to reacquire it after. This way it's clear that the lock is released (when looking only at a callsite). This pattern has one drawback though: it may introduce unnecessary lock operations.